### PR TITLE
Remove `ArgumentError()` in `parse_cache_header()` when `@depot` cannot be resolved

### PR DESCRIPTION
**Pull request by staticfloat**
In the original implementation of relocatable package cache files, `parse_cache_header()` was made responsible for resolving `@depot/`-relative paths, and it threw an `ArgumentError()` if it could not find one.  However, this could happen for a few reasons:

1. The `JULIA_DEPOT_PATH` could be set incorrectly, so the appropriate source text files could not be found in an appropriate `packages/` directory.

2. The package could have been upgraded, leaving a cache file for an older version on-disk, but no matching source files.

For a concrete example of (2), see the following `bash` script:

    #!/bin/bash
    set -euo pipefail

    TEMP_PATH=$(mktemp -d)
    JULIA=${JULIA:-julia}
    echo JULIA: ${JULIA}
    export JULIA_DEPOT_PATH="${TEMP_PATH}"
    trap 'rm -rf ${TEMP_PATH}' EXIT

    # Create a `Foo.jl` that has an `internal.jl` within it
    FOO_DIR=${JULIA_DEPOT_PATH}/dev/Foo
    mkdir -p ${FOO_DIR}/src
    cat >${FOO_DIR}/Project.toml <<EOF
    name = "Foo"
    uuid = "00000000-0000-0000-0000-000000000001"
    version = "1.0.0"
    EOF
    cat >${FOO_DIR}/src/Foo.jl <<EOF
    module Foo

    include("internal.jl")
    end
    EOF

    cat >${FOO_DIR}/src/internal.jl <<EOF
    # Nothing to see here, folks
    EOF

    ${JULIA} -e "import Pkg; Pkg.develop(\"Foo\"); Pkg.precompile()"

    # Change `Foo` to no longer depend on `internal.jl`; this should invalidate its cache files
    cat >${FOO_DIR}/src/Foo.jl <<EOF
    module Foo
    end
    EOF
    rm -f ${FOO_DIR}/src/internal.jl

    # This should print `SUCCESS!`, not `FAILURE!`
    ${JULIA} -e 'using Foo; println("SUCCESS!")' || echo "FAILURE!"

This pull request removes the `ArgumentError()`, as it seems reasonable to require `parse_cache_header()` to not throw errors in cases like these.  We instead rely upon a higher level validation to reject a cachefile whose depot cannot be found, which should happen when `stale_cachefile()` later checks to ensure that all includes are found on-disk, (which will be false, as these paths start with `@depot/`, an unlikely path prefix to exist).

### CODE REVIEW

**Comment by staticfloat** [comment]
Added a test

**Review by vchuravy — APPROVED** [review]

**Review by vtjnash — APPROVED** [review]

File: test/loading.jl [hunk 1]
```diff
@@ -1274,3 +1274,60 @@ end
...
+
+        # In our depot, `dev` and then `precompile` this `Foo` package.
+        @test success(addenv(
+            `$(Base.julia_cmd()) --startup-file=no -e 'import Pkg; Pkg.develop("Foo"); Pkg.precompile(); exit(0)'`,
+            "JULIA_DEPOT_PATH" => depot,
+        ))
+
+        # Get the size of the generated `.ji` file so that we can ensure that it gets altered
+        foo_compiled_path = joinpath(depot, "compiled", "v$(VERSION.major).$(VERSION.minor)", "Foo")
+        cache_path = joinpath(foo_compiled_path, only(filter(endswith(".ji"), readdir(foo_compiled_path))))
```
Comment (fatteneder) [inline test/loading.jl:1309]: Just a minor thing: One could push `depot` onto the LOAD_PATH and use something like (untested)
```
pkg = Base.identify_package("Foo")
Base.compilecache(pkg)
```
to get all `.ji` files. Same can be done below.

**Review by fatteneder — COMMENTED** [review]
This looks good.

And I think this solves the problem of silent PkgEval skips which I discovered yesterday in https://github.com/JuliaLang/julia/pull/51746#issuecomment-1788140104.
[ But it does not address the real problem there which IIUC is the use of `dep/build.jl` in the pkgs. ]

File: test/loading.jl (hunk 1 above)
Comment (staticfloat) [reply test/loading.jl:1309]: I didn't feel this was important enough to justify waiting another 24 hours for CI to catch up; if you feel strongly about it, feel free to open another PR.

**Pull request merged**